<a href="https://colab.research.google.com/github/ByeolHan/ByeolHan/blob/main/CNN_%EB%AA%A8%EB%8D%B8_%EC%84%B1%EB%8A%A5%ED%8F%89%EA%B0%80.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 데이터 전처리

In [1]:
import os, sys 
from google.colab import drive 
drive.mount('/content/mnt') 
nb_path = '/content/notebooks' 
os.symlink('/content/mnt/My Drive/Colab Notebooks', nb_path) 
sys.path.insert(0, nb_path)

Mounted at /content/mnt


In [2]:
from PIL import Image
import matplotlib
import matplotlib.pyplot as plt
import torch
from torch.utils.data import Dataset, DataLoader
import numpy as np
import torchvision
from torchvision import transforms

In [3]:
preprocess = transforms.Compose([
    transforms.Resize(256), # 사이즈 처리
    transforms.CenterCrop(224), # 화면에 여백이 없도록 사이즈 조정
    transforms.ToTensor(), # 이미지를 텐서형태로 변환
    transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]), # 정규화
])

test_set = torchvision.datasets.ImageFolder(root='/content/notebooks/ILSVRC2012_img_val',
                                            transform = preprocess)

In [4]:
# 배치형식으로 네트워크에 올릴 수 있게 준비
testloader = DataLoader(test_set, batch_size=64,shuffle=False, num_workers=2)

In [ ]:
# 확인차
dataiter = iter(testloader)
images, labels = dataiter.next()
print(labels)

# 모델 불러오기 / 성능 확인

## 1. Alexnet

In [ ]:
import torch
import torchvision
model = torch.hub.load('pytorch/vision:v0.9.0', 'alexnet', pretrained=True)
model.eval()

In [ ]:
# 성능 함수
correct_Alexnet = 0
total_Alexnet = 0

if torch.cuda.is_available():
    model.to('cuda')

with torch.no_grad():
    for data in testloader:      
        images, labels = data
        images = images.to('cuda')   
        labels = labels.to('cuda')    
        outputs = model(images)
        _, predicted = torch.max(outputs.data, 1) #top-1 accuracy
        total_Alexnet += labels.size(0)
        correct_Alexnet += (predicted == labels).sum().item()

print('Accuracy of the network on the 50000 test images: %d %%' % (
    100 * correct_Alexnet / total_Alexnet))

Accuracy of the network on the 50000 test images: 56 %


## 2. VGG16

In [ ]:
import torch
import torchvision
model = torch.hub.load('pytorch/vision:v0.9.0', 'vgg16', pretrained=True)
model.eval()

In [6]:
# 성능 함수
correct_VGG16 = 0
total_VGG16 = 0

if torch.cuda.is_available():
    model.to('cuda')

with torch.no_grad():
    for data in testloader:      
        images, labels = data
        images = images.to('cuda')   
        labels = labels.to('cuda')    
        outputs = model(images)
        _, predicted = torch.max(outputs.data, 1) #top-1 accuracy
        total_VGG16 += labels.size(0)
        correct_VGG16 += (predicted == labels).sum().item()

print('Accuracy of the network on the 50000 test images: %d %%' % (
    100 * correct_VGG16 / total_VGG16))

Accuracy of the network on the 50000 test images: 71 %


## 3. ResNet18

In [ ]:
# resnet18 모델 로딩
import torch
import torchvision
model = torch.hub.load('pytorch/vision:v0.9.0', 'resnet18', pretrained=True)
model.eval()

In [8]:
# 성능 함수
correct_ResNet = 0
total_ResNet = 0

if torch.cuda.is_available():
    model.to('cuda')

with torch.no_grad():
    for data in testloader:      
        images, labels = data
        images = images.to('cuda')   
        labels = labels.to('cuda')    
        outputs = model(images)
        _, predicted = torch.max(outputs.data, 1) #top-1 accuracy
        total_ResNet += labels.size(0)
        correct_ResNet += (predicted == labels).sum().item()

print('Accuracy of the network on the 50000 test images: %d %%' % (
    100 * correct_ResNet / total_ResNet))

Accuracy of the network on the 50000 test images: 69 %


## 4. GoogLeNet

In [ ]:
import torch
import torchvision
model = torch.hub.load('pytorch/vision:v0.9.0', 'googlenet', pretrained=True)
model.eval()

In [10]:
# 성능 함수
correct_GoogLeNet = 0
total_GoogLeNet = 0

if torch.cuda.is_available():
    model.to('cuda')

with torch.no_grad():
    for data in testloader:      
        images, labels = data
        images = images.to('cuda')   
        labels = labels.to('cuda')    
        outputs = model(images)
        _, predicted = torch.max(outputs.data, 1) #top-1 accuracy
        total_GoogLeNet += labels.size(0)
        correct_GoogLeNet += (predicted == labels).sum().item()

print('Accuracy of the network on the 50000 test images: %d %%' % (
    100 * correct_GoogLeNet / total_GoogLeNet))

Accuracy of the network on the 50000 test images: 69 %


# 성능 비교

|모델|AlexNet|VGG16|ResNet|GoogLeNet|
|---|---|---|---|---|
|성능|56%|71%|69%|69%|
